File: test/e2e/Apps/BasicPython/entity_error_handling.py
#
# Copyright (c) Microsoft. All rights reserved.
# Licensed under the MIT license. See LICENSE file in the project root for full license information.
#

import azure.durable_functions as df

from custom_exception_types import InvalidOperationException, OverflowException

bp = df.Blueprint()

attempt_count = {}


@bp.orchestration_trigger(context_name="context", orchestration="ThrowEntityOrchestration")
def rethrow_entity_exception(context: df.DurableOrchestrationContext):
    entityId = df.EntityId("Counter", "myCounter")
    _ = yield context.call_entity(entityId, "get", context.instance_id)
    return "Success"


@bp.orchestration_trigger(context_name="context", orchestration="CatchEntityOrchestration")
def catch_entity_exception(context: df.DurableOrchestrationContext):
    entityId = df.EntityId("Counter", "myCounter")
    try:
        _ = yield context.call_entity(entityId, "get", context.instance_id)
        return "Success"
    except Exception as e:
        return str(e)


@bp.orchestration_trigger(context_name="context", orchestration="RetryEntityOrchestration")
def retry_entity_function(context: df.DurableOrchestrationContext):
    entityId = df.EntityId("Counter", "myCounter")
    try:
        _ = yield context.call_entity(entityId, "get", context.instance_id)
        return "Success"
    except Exception as e:
        _ = yield context.call_entity(entityId, "get", context.instance_id)
        return "Success"


@bp.entity_trigger(context_name="context")
def Counter(context):
    global attempt_count
    instance_id = context.get_input()
    if not instance_id:
        raise "Did not get a valid instanceId as input to the entity"
    if instance_id not in attempt_count:
        attempt_count[instance_id] = 1
        raise InvalidOperationException(
            "This entity failed\r\nMore information about the failure") from OverflowException("Inner exception message")
    attempt_count[instance_id] += 1
    context.set_result(0)

File: test/e2e/Apps/BasicNode/src/functions/EntityErrorHandling.ts
// Copyright (c) .NET Foundation. All rights reserved.
// Licensed under the MIT License. See License.txt in the project root for license information.

import * as df from "durable-functions";
import { OrchestrationContext, OrchestrationHandler, EntityContext, EntityHandler } from "durable-functions";
import { InvalidOperationException, OverflowException } from "../Shared/ExceptionTypes";

// In-memory attempt count (not durable, but matches Python's global dict for test purposes)
const attemptCount: Record<string, number> = {};

// Orchestration: ThrowEntityOrchestration
const ThrowEntityOrchestration: OrchestrationHandler = function* (context: OrchestrationContext) {
    const entityId = new df.EntityId("Counter", "myCounter");
    yield context.df.callEntity(entityId, "get", context.df.instanceId);
    return "Success";
};
df.app.orchestration("ThrowEntityOrchestration", ThrowEntityOrchestration);

// Orchestration: CatchEntityOrchestration
const CatchEntityOrchestration: OrchestrationHandler = function* (context: OrchestrationContext) {
    const entityId = new df.EntityId("Counter", "myCounter");
    try {
        yield context.df.callEntity(entityId, "get", context.df.instanceId);
        return "Success";
    } catch (e: any) {
        return String(e);
    }
};
df.app.orchestration("CatchEntityOrchestration", CatchEntityOrchestration);

// Orchestration: RetryEntityOrchestration
const RetryEntityOrchestration: OrchestrationHandler = function* (context: OrchestrationContext) {
    const entityId = new df.EntityId("Counter", "myCounter");
    try {
        yield context.df.callEntity(entityId, "get", context.df.instanceId);
        return "Success";
    } catch (e) {
        yield context.df.callEntity(entityId, "get", context.df.instanceId);
        return "Success";
    }
};
df.app.orchestration("RetryEntityOrchestration", RetryEntityOrchestration);

// Entity: Counter
const Counter: EntityHandler<string> = (context: EntityContext<string>) => {
    const instanceId = context.df.getInput<string>();
    if (!instanceId) {
        throw new Error("Did not get a valid instanceId as input to the entity");
    }
    if (!(instanceId in attemptCount)) {
        attemptCount[instanceId] = 1;
        const inner = new OverflowException("Inner exception message");
        const err = new InvalidOperationException("This entity failed\r\nMore information about the failure");
        err.cause = inner;
        throw err;
    }
    attemptCount[instanceId] += 1;
    context.df.return(0);
};
df.app.entity("Counter", Counter);